### PR TITLE
[INV-58] Create an endpoint to pay for a booking using Stripe

### PR DESCRIPTION
**Pull request by ShahobiddinIsmoilov**
### Demonstration

https://github.com/user-attachments/assets/5ffd1f2e-6752-4110-b397-2183be37a02e

### Tests
Screenshot 2024-11-27 213646

### CODE REVIEW

File: src/modules/payments/services/payments.service.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    const paymentRecord = existingPaymentRecord
+      ? existingPaymentRecord
+      : newPaymentRecord;
+
+    paymentRecord.status = status;
+    await this.paymentRepository.save(paymentRecord);
+  }
+
+  async getPaymentRecordByBookingId(bookingId: string) {
+    const paymentRecord = this.paymentRepository.findOneBy({ bookingId });
```
Comment (farrooh) [inline src/modules/payments/services/payments.service.ts:37]: Will a new transaction be created if one is failed? 
As I understand from your code in payment.service, the answer is no, so you must add a unique constraint for bookingId to the payment table in DB. 
But if yes, there might be multiple transactions for a single booking, then you should add status checking too here.

File: src/modules/payments/services/payments.service.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    const newPaymentRecord = this.paymentRepository.create({
+      amount,
+      transactionId,
+      bookingId,
+    });
+
+    const paymentRecord = existingPaymentRecord
+      ? existingPaymentRecord
+      : newPaymentRecord;
```
Comment (farrooh) [inline src/modules/payments/services/payments.service.ts:30]: Or just say `= existingPaymentRecord || newPaymentRecord;`

**Review by farrooh — CHANGES_REQUESTED** [review]
1) Add a unique constraint for bookingID in the payment table if there won't be several transactions for a single booking.
2) Make it impossible to delete completed (already made payment) bookings. Their userID should be set to null if the user is deleted. And make bookingID not null in the payment table. Payments must have an entity.

File: src/modules/payments/services/stripe.service.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import { PaymentStatus } from '#/shared/constants/payment-status.constant';
+import { buildStripeLineItems } from '../helpers/stripe.helper';
+import { StripeConfig } from '../types/stripe-config.type';
+import { PaymentsService } from './payments.service';
+import { BookingsService } from '#/modules/bookings/services/bookings.service';
+import { BookingStatus } from '#/shared/constants/booking-status.constant';
+import { AccommodationService } from '#/modules/accommodations/services/accommodations.service';
+
+@Injectable()
+export class StripeService {
```
Comment (NegatiV-64) [inline src/modules/payments/services/stripe.service.ts:22]: I'd call it CheckoutService

File: src/modules/payments/services/stripe.service.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  ) {
+    const { bookingId, items } = stripeCreateCheckoutReqDto;
+
+    const existingBooking = await this.bookingsService.getBookingById(
+      bookingId,
+      accountId,
+    );
+
+    if (!existingBooking) {
+      throw new BadRequestException('Booking does not exist');
```
[the remark this answers is not in the document]
Comment (NegatiV-64) [reply src/modules/payments/services/stripe.service.ts:43]: I think that BadRequest could make sense

File: src/modules/payments/services/stripe.service.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  async handleWebhook(rawBody: Buffer, signature: string) {
+    const event = this.stripe.webhooks.constructEvent(
+      rawBody,
+      signature,
+      this.stripeConfig.webhookSecret,
+    );
+
+    let paymentStatus: PaymentStatus = PaymentStatus.Pending;
+
+    switch (event.type) {
```
Comment (NegatiV-64) [inline src/modules/payments/services/stripe.service.ts:84]: I think that map could be a better solution

File: src/shared/extractors/payment-intent.extractor.ts
Comment (NegatiV-64) [inline src/shared/extractors/payment-intent.extractor.ts:1]: Move it to the payment module, why it is on the shared level?

File: src/shared/utils/cors.util.ts
Comment (NegatiV-64) [inline src/shared/utils/cors.util.ts:1]: It should be inside the config, it is not a real utility

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]

**Comment by Savesh1508** [comment]
Bro, in the next time, let's do huge renamings in the another pr. som many files changed so it's painful to check what I want

**Review by NegatiV-64 — APPROVED** [review]

**Pull request merged**